File: 20-actors/fuchi/methods/test_couple.py
#!/usr/bin/env python3
"""Tests for 扶持 (fuchi) couple.py — R1(d) Displacement-Dividend cohort coupling.

Standalone-runnable: python3 test_couple.py
"""
from __future__ import annotations

import sys

from couple import (
    TITHE_BPS,
    CohortEarmark,
    DisplacementEvent,
    coupling_gate,
    earmark_from_surplus,
    events_from_seed,
)


def _event(**over):
    base = dict(displacing_actor="sanae", cohort_id="cohort-sanae",
                displaced_count=12, surplus_usd_micros_yr=60_000_000_000, funded=True)
    base.update(over)
    return DisplacementEvent(**base)


# ── TitheRouter split ───────────────────────────────────────────────────────
def test_tithe_split_is_ten_percent_exact():
    e = earmark_from_surplus(_event(surplus_usd_micros_yr=60_000_000_000))
    assert e.tithe_usd_micros == 6_000_000_000
    assert e.earmark_usd_micros_yr == 54_000_000_000


def test_gross_equals_tithe_plus_earmark_always():
    for s in (1, 7, 999, 10_001, 60_000_000_000, 123_456_789):
        e = earmark_from_surplus(_event(surplus_usd_micros_yr=s))
        assert e.tithe_usd_micros + e.earmark_usd_micros_yr == s


def test_tithe_bps_is_1000():
    assert TITHE_BPS == 1000


def test_earmark_construction_rejects_inexact_split():
    try:
        CohortEarmark("c", "sanae", 100, 5, 90, True)  # 5+90 != 100
    except ValueError as e:
        assert "split" in str(e).lower()
        return
    raise AssertionError("inexact gross/tithe/earmark split must be refused")


# ── G2 coupling gate ────────────────────────────────────────────────────────
def test_funded_cohort_within_earmark_is_admissible():
    e = _event(funded=True)
    em = earmark_from_surplus(e)
    g = coupling_gate(e, em, committed_floor_usd_micros_yr=20_000_000_000)
    assert g["admissible"] is True and g["headroom"] > 0


def test_unfunded_cohort_is_refused():
    e = _event(funded=False)
    em = earmark_from_surplus(e)
    g = coupling_gate(e, em, committed_floor_usd_micros_yr=1)
    assert g["admissible"] is False and "G2" in g["reason"] and "no funded" in g["reason"]


def test_over_earmark_commitment_is_refused():
    e = _event(surplus_usd_micros_yr=10_000_000_000, funded=True)  # earmark 9_000_000_000
    em = earmark_from_surplus(e)
    g = coupling_gate(e, em, committed_floor_usd_micros_yr=20_000_000_000)
    assert g["admissible"] is False and "exceeds funded earmark" in g["reason"]


def test_committed_exactly_at_earmark_is_admissible():
    e = _event(surplus_usd_micros_yr=10_000_000_000, funded=True)  # earmark 9_000_000_000
    em = earmark_from_surplus(e)
    g = coupling_gate(e, em, committed_floor_usd_micros_yr=9_000_000_000)
    assert g["admissible"] is True and g["headroom"] == 0


def test_negative_surplus_refused():
    try:
        DisplacementEvent("sanae", "c", 1, -5)
    except ValueError:
        return
    raise AssertionError("negative surplus must be refused")


# ── seed parsing ────────────────────────────────────────────────────────────
def test_events_from_seed_reads_funded_flag():
    recs = [{":event/displacing-actor": "sanae", ":event/cohort-id": "c1",
             ":event/displaced-count": 12, ":event/surplus-usd-micros-yr": 60_000_000_000,
             ":event/funded": True},
            {":event/displacing-actor": "hataori", ":event/cohort-id": "c2",
             ":event/displaced-count": 30, ":event/surplus-usd-micros-yr": 0,
             ":event/funded": False}]
    evs = events_from_seed(recs)
    assert evs[0].funded is True and evs[1].funded is False


def _run():
    fns = [v for k, v in sorted(globals().items()) if k.startswith("test_") and callable(v)]
    for fn in fns:
        fn()
    print(f"test_couple.py: {len(fns)} passed")
    return 0


if __name__ == "__main__":
    sys.exit(_run())

File: 20-actors/fuchi/methods/test_couple.clj
;; test_couple.clj — 扶持 couple: Displacement-Dividend coupling (TitheRouter split + G2 gate),
;; parity with couple.py. Run via `bb test:fuchi`. ADR-2606142300.
(ns fuchi.methods.test-couple
  (:require [clojure.test :refer [deftest is testing run-tests]]
            [fuchi.methods.couple :as c]
            [fuchi.methods.live-gate :as lg]))

(def ^:private funded-event
  (c/make-displacement-event {:displacing-actor "sanae" :cohort-id "c1" :displaced-count 10
                              :surplus-usd-micros-yr 100000000 :funded true}))
(def ^:private earmark (c/earmark-from-surplus funded-event))

(deftest tithe-split-exact
  (testing "10% TitheRouter split is an exact integer split (gross = tithe + earmark) — golden"
    (is (= 100000000 (:gross-usd-micros-yr earmark)))
    (is (= 10000000 (:tithe-usd-micros earmark)))
    (is (= 90000000 (:earmark-usd-micros-yr earmark)))
    ;; odd surplus — no rounding leak
    (let [e (c/earmark-from-surplus (c/make-displacement-event {:displacing-actor "a" :cohort-id "c"
                                                                :surplus-usd-micros-yr 12345 :funded true}))]
      (is (= 1234 (:tithe-usd-micros e)))
      (is (= 11111 (:earmark-usd-micros-yr e)))
      (is (= 12345 (+ (:tithe-usd-micros e) (:earmark-usd-micros-yr e)))))
    (is (thrown? Exception (c/make-cohort-earmark {:gross-usd-micros-yr 100 :tithe-usd-micros 10
                                                   :earmark-usd-micros-yr 80})))))  ; 10+80≠100

(deftest g2-coupling-gate
  (testing "G2 — admissible iff funded AND committed ≤ earmark (golden from couple.py)"
    (let [ok (c/coupling-gate funded-event earmark 50000000)]
      (is (true? (:admissible ok)))
      (is (= 40000000 (:headroom ok))))                      ; 90M − 50M
    (is (false? (:admissible (c/coupling-gate funded-event earmark 95000000))))   ; over earmark
    (let [unfunded (c/earmark-from-surplus (c/make-displacement-event
                                            {:displacing-actor "x" :cohort-id "c" :surplus-usd-micros-yr 100000000 :funded false}))]
      (is (false? (:admissible (c/coupling-gate funded-event unfunded 1))))        ; not funded
      (is (= 0 (:headroom (c/coupling-gate funded-event unfunded 1)))))))

(deftest event-guards
  (testing "displacement event input guards"
    (is (thrown? Exception (c/make-displacement-event {:surplus-usd-micros-yr -1})))
    (is (thrown? Exception (c/make-displacement-event {:displaced-count -1})))))

(deftest commit-live-stacks-two-refusals
  (testing "commit-live: R2 gate + G2 coupling both apply (no live displacement without funded cohort)"
    (let [g  (lg/make-gate {:leg "couple"})
          cm (c/commit-live funded-event earmark 50000000 g)]
      (is (true? (:admissible cm)))
      (is (= 7 (:council-level cm)))                          ; couple leg = Lv7
      (is (= 50000000 (:committed-usd-micros-yr cm)))
      ;; G2 refusal: unfunded cohort cannot commit even through the gate
      (let [unfunded (c/earmark-from-surplus (c/make-displacement-event
                                              {:displacing-actor "x" :cohort-id "c" :surplus-usd-micros-yr 100000000 :funded false}))]
        (is (thrown? Exception (c/commit-live funded-event unfunded 1 g))))
      ;; G2 refusal: over-committed beyond the funded earmark
      (is (thrown? Exception (c/commit-live funded-event earmark 95000000 g))))))

(when (= *file* (System/getProperty "babashka.file"))
  (let [{:keys [fail error]} (run-tests 'fuchi.methods.test-couple)]
    (System/exit (if (pos? (+ fail error)) 1 0))))
